policies.go

```diff
@@ -620,6 +620,244 @@ func (t *tokenAwareHostPolicy) Pick(qry ExecutableQuery) NextHost {
+	// reads can be unlocked as long as they are not used for updating state later.
+	mu          sync.Mutex
+	hosts       cowHostList
+	partitioner string
+	metadata    atomic.Value // *clusterMeta
+
+	session *Session
+	logger  StdLogger
+}
+
+func (p *ybPartitionAwareHostPolicy) KeyspaceChanged(update KeyspaceUpdateEvent) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	meta := p.getMetadataForUpdate()
+	p.updateReplicas(meta, update.Keyspace)
+	p.metadata.Store(meta)
+}
+
+func (p *ybPartitionAwareHostPolicy) Init(s *Session) {
+	p.getKeyspaceMetadata = s.KeyspaceMetadata
+	p.getKeyspaceName = func() string { return s.cfg.Keyspace }
+	p.session = s
+	p.logger = s.logger
+}
+
+func (p *ybPartitionAwareHostPolicy) IsLocal(host *HostInfo) bool {
+	return p.fallback.IsLocal(host)
+}
+
+func (p *ybPartitionAwareHostPolicy) getMetadataReadOnly() *clusterMeta {
+	meta, _ := p.metadata.Load().(*clusterMeta)
+	return meta
+}
+
+// getMetadataForUpdate returns clusterMeta suitable for updating.
+// It is a SHALLOW copy of current metadata in case it was already set or new empty clusterMeta otherwise.
+// This function should be called with t.mu mutex locked and the mutex should not be released before
+// storing the new metadata.
+func (p *ybPartitionAwareHostPolicy) getMetadataForUpdate() *clusterMeta {
+	metaReadOnly := p.getMetadataReadOnly()
+	meta := new(clusterMeta)
+	if metaReadOnly != nil {
+		*meta = *metaReadOnly
+	}
+	return meta
+}
+
+func (p *ybPartitionAwareHostPolicy) updateReplicas(meta *clusterMeta, keyspace string) {
+	newReplicas := make(map[string]tokenRingReplicas, len(meta.replicas))
+
+	ks, err := p.getKeyspaceMetadata(keyspace)
+	if err == nil {
+		strat := getStrategy(ks, p.logger)
+		if strat != nil {
+			if meta != nil && meta.tokenRing != nil {
+				newReplicas[keyspace] = strat.replicaMap(meta.tokenRing)
+			}
+		}
+	}
+
+	for ks, replicas := range meta.replicas {
+		if ks != keyspace {
+			newReplicas[ks] = replicas
+		}
+	}
+
+	meta.replicas = newReplicas
+}
+
+func (p *ybPartitionAwareHostPolicy) AddHost(host *HostInfo) {
+	p.mu.Lock()
+	if p.hosts.add(host) {
+		meta := p.getMetadataForUpdate()
+		meta.resetTokenRing(p.partitioner, p.hosts.get(), p.logger)
+		p.updateReplicas(meta, p.getKeyspaceName())
+		p.metadata.Store(meta)
+	}
+	p.mu.Unlock()
+
+	p.fallback.AddHost(host)
+}
+
+func (p *ybPartitionAwareHostPolicy) HostUp(host *HostInfo) {
+	p.fallback.HostUp(host)
+}
+
+func (p *ybPartitionAwareHostPolicy) HostDown(host *HostInfo) {
+	p.fallback.HostDown(host)
+}
+
+func (p *ybPartitionAwareHostPolicy) RemoveHost(host *HostInfo) {
+	p.mu.Lock()
+	if p.hosts.remove(host.ConnectAddress()) {
+		meta := p.getMetadataForUpdate()
+		meta.resetTokenRing(p.partitioner, p.hosts.get(), p.logger)
+		p.updateReplicas(meta, p.getKeyspaceName())
+		p.metadata.Store(meta)
+	}
+	p.mu.Unlock()
+
+	p.fallback.RemoveHost(host)
+}
+
+func (p *ybPartitionAwareHostPolicy) SetPartitioner(partitioner string) {
```

Comment:
Is this required for our policy? Please remove, if not. Similarly, check other functions.

Reply:
All policies extend\return [HostSelectionPolicy](https://github.com/yugabyte/gocql/blob/partition_aware_policy/policies.go#L265) interface so each policy needs to have a implementation of all the functions that are part of the interface due to which we cannot remove such functions but since we do not need these functions for PartitionAwarePolicy they are empty implementations for YBPartitionAwarePolicy.